src/dbnode/encoding/m3tsz/iterator.go

```diff
@@ -384,30 +225,365 @@ func (it *readerIterator) hasNext() bool {
+		if err != nil {
+			return 0, false, err
+		}
+		markerOrDOD, err := it.readMarkerOrDeltaOfDelta(stream)
+		if err != nil {
+			return 0, false, err
+		}
+		return markerOrDOD, true, nil
+	case mes.TimeUnit():
+		_, err := stream.ReadBits(numBits)
+		if err != nil {
+			return 0, false, err
+		}
+		err = it.ReadTimeUnit(stream)
+		if err != nil {
+			return 0, false, err
+		}
+		markerOrDOD, err := it.readMarkerOrDeltaOfDelta(stream)
+		if err != nil {
+			return 0, false, err
+		}
+		return markerOrDOD, true, nil
+	default:
+		return 0, false, nil
+	}
+}
+
+func (it *TimestampIterator) readMarkerOrDeltaOfDelta(stream encoding.IStream) (time.Duration, error) {
+	if !it.SkipMarkers {
+		dod, success, err := it.tryReadMarker(stream)
+		if err != nil {
+			return 0, err
+		}
+		if it.Done {
+			return 0, nil
+		}
+
+		if success {
+			return dod, nil
+		}
+	}
+
+	tes, exists := it.Opts.TimeEncodingSchemes()[it.TimeUnit]
+	if !exists {
+		return 0, fmt.Errorf("time encoding scheme for time unit %v doesn't exist", it.TimeUnit)
+	}
+
+	return it.readDeltaOfDelta(stream, tes)
+}
+
+func (it *TimestampIterator) readDeltaOfDelta(
+	stream encoding.IStream, tes encoding.TimeEncodingScheme) (time.Duration, error) {
+	if it.TimeUnitChanged {
+		// NB(xichen): if the time unit has changed, always read 64 bits as normalized
+		// dod in nanoseconds.
+		dodBits, err := stream.ReadBits(64)
+		if err != nil {
+			return 0, err
+		}
+
+		dod := encoding.SignExtend(dodBits, 64)
+		return time.Duration(dod), nil
+	}
+
+	cb, err := stream.ReadBits(1)
+	if err != nil {
+		return 0, err
+	}
+	if cb == tes.ZeroBucket().Opcode() {
+		return 0, nil
+	}
+
+	buckets := tes.Buckets()
+	for i := 0; i < len(buckets); i++ {
+		nextCB, err := stream.ReadBits(1)
+		if err != nil {
+			return 0, nil
+		}
+
+		cb = (cb << 1) | nextCB
+		if cb == buckets[i].Opcode() {
+			dodBits, err := stream.ReadBits(buckets[i].NumValueBits())
+			if err != nil {
+				return 0, err
+			}
+
+			dod := encoding.SignExtend(dodBits, buckets[i].NumValueBits())
+			timeUnit, err := it.TimeUnit.Value()
+			if err != nil {
+				return 0, nil
+			}
+
+			return xtime.FromNormalizedDuration(dod, timeUnit), nil
+		}
+	}
+
+	numValueBits := tes.DefaultBucket().NumValueBits()
+	dodBits, err := stream.ReadBits(numValueBits)
+	if err != nil {
+		return 0, err
+	}
+
+	dod := encoding.SignExtend(dodBits, numValueBits)
+	timeUnit, err := it.TimeUnit.Value()
+	if err != nil {
+		return 0, nil
+	}
+
+	return xtime.FromNormalizedDuration(dod, timeUnit), nil
+}
+
+func (it *TimestampIterator) readAnnotation(stream encoding.IStream) error {
+	antLen, err := it.readVarint(stream)
+	if err != nil {
+		return err
+	}
+
+	// NB: we add 1 here to offset the 1 we subtracted during encoding.
+	antLen = antLen + 1
+	if antLen <= 0 {
+		return fmt.Errorf("unexpected annotation length %d", antLen)
+	}
+
+	// TODO(xichen): use pool to allocate the buffer once the pool diff lands.
+	buf := make([]byte, antLen)
```

Comment:
Hm, this always allocated huh?  Pretty bad lol.  

Ah, is it worth just allocating to `PrevAnt` if the capacity of that is not at `antLen` and then read directly into that?  (Not sure if that would effect the read path, but if this is only for internal tracking then there's no reason to allocate here each time)

Reply:
I can try and address in my follow up. Don't want to mess with this right now haha. Doesn't affect any of our workloads right now cause its only if an annotation exists which we're not really using, although could be helpful for the index agreed